src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerListener.cs

```diff
@@ -54,21 +62,49 @@ public void Dispose()
 
         public async Task StartAsync(CancellationToken cancellationToken)
         {
+            if (this.listenerStarted)
+            {
+                throw new InvalidOperationException("The listener has already been started.");
+            }
+
+            this.InitializeHost();
+
             try
             {
-                await host.RegisterObserverFactoryAsync(this);
+                await this.host.RegisterObserverFactoryAsync(this);
+                this.listenerStarted = true;
             }
             catch (DocumentClientException ex) when (ex.StatusCode == HttpStatusCode.NotFound)
             {
                 // Throw a custom error so that it's easier to decipher.
                 string message = $"Either the source collection '{monitorCollection.CollectionName}' (in database '{monitorCollection.DatabaseName}')  or the lease collection '{leaseCollection.CollectionName}' (in database '{leaseCollection.DatabaseName}') does not exist. Both collections must exist before the listener starts. To automatically create the lease collection, set '{nameof(CosmosDBTriggerAttribute.CreateLeaseCollectionIfNotExists)}' to 'true'.";
+                this.host = null;
                 throw new InvalidOperationException(message, ex);
             }
         }
 
-        public Task StopAsync(CancellationToken cancellationToken)
+        public async Task StopAsync(CancellationToken cancellationToken)
         {
-            return host.UnregisterObserversAsync();
+            try
+            {
+                if (this.host != null && this.listenerStarted)
+                {
+                    await this.host.UnregisterObserversAsync();
+                    this.listenerStarted = false;
+                }
+            }
+            catch (Exception ex)
+            {
+                this.logger.LogWarning($"Stopping the observer failed, potentially it was never started. Exception: {ex.Message}.");
```

Comment:
It looks the logger instance passed into this class may be null (based on `_logger = loggerFactory?.CreateLogger`). Since you require a non-null logger instance here, recommend ensuring it can't be null upstream.

Reply:
Good catch!